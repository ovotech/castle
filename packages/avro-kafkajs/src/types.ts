import {
  EachMessagePayload,
  Batch,
  EachBatchPayload,
  Message,
  ProducerRecord,
  TopicMessages,
  ProducerBatch,
  KafkaMessage,
  ConsumerRunConfig,
} from 'kafkajs';
import { Schema } from 'avsc';

export interface AvroEachMessagePayload<T = unknown> extends Omit<EachMessagePayload, 'message'> {
  message: AvroKafkaMessage<T>;
}

export interface AvroBatch<T = unknown> extends Omit<Batch, 'messages'> {
  messages: AvroKafkaMessage<T>[];
}

export interface AvroEachBatchPayload<T = unknown> extends Omit<EachBatchPayload, 'batch'> {
  batch: AvroBatch<T>;
}

export type AvroEachMessage<T = unknown> = (payload: AvroEachMessagePayload<T>) => Promise<void>;
export type AvroEachBatch<T = unknown> = (payload: AvroEachBatchPayload<T>) => Promise<void>;

export interface AvroKafkaMessage<T = unknown> extends Omit<KafkaMessage, 'value'> {
  schema: Schema;
  value: T;
}

export interface AvroMessage<T = unknown> extends Omit<Message, 'value'> {
  value: T;
}

export interface AvroProducerRecord<T = unknown> extends Omit<ProducerRecord, 'messages'> {
  schema: Schema;
  messages: AvroMessage<T>[];
}

export interface AvroTopicMessages<T = unknown> extends Omit<TopicMessages, 'messages'> {
  schema: Schema;
  messages: AvroMessage<T>[];
}

export interface AvroProducerBatch extends Omit<ProducerBatch, 'topicMessages'> {
  topicMessages: AvroTopicMessages[];
}

export interface AvroConsumerRun<T = unknown>
  extends Omit<ConsumerRunConfig, 'eachBatch' | 'eachMessage'> {
  eachBatch?: (payload: AvroEachBatchPayload<T>) => Promise<void>;
  eachMessage?: (payload: AvroEachMessagePayload<T>) => Promise<void>;
}

export type TopicsAlias = { [key: string]: string };
