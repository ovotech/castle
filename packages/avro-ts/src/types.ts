import * as ts from 'typescript';
import { Document, DocumentContext } from '@ovotech/ts-compose';
import { Schema } from 'avsc';

export type CustomLogicalType =
  | { module: string; named: string }
  | { module: string; defaultAs: string }
  | { module: string; allAs: string };

export interface Context extends DocumentContext {
  logicalTypes?: { [key: string]: string | CustomLogicalType };
  namespace?: string;
  refs?: { [key: string]: Schema };
  external?: { [file: string]: { [key: string]: Schema } };
  defaultsAsOptional?: boolean;
}

export type Convert<TSchema = Schema, TType = ts.TypeNode> = (
  context: Context,
  type: TSchema,
) => Document<TType, Context>;
