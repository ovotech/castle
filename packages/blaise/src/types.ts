import { AvroKafkaMessage } from '@ovotech/avro-kafkajs/dist/types';
import { Options as AvroOptions, AvroMock } from '@ovotech/avro-mock-generator';
import {
  CastleEachBatchPayload,
  CastleEachMessagePayload,
} from '@ovotech/castle';
import { schema } from 'avsc';
import { DeepOmit, DeepPartial } from 'ts-essentials';

export interface Blaise<T = unknown> {
  <T1 = T>(newDefaults?: DeepPartial<BlaiseDefaults<T>>): Blaise<T1>;
  default<T1 = T>(newDefaults: DeepPartial<BlaiseDefaults<T>>): Blaise<T1>;
  pickUnion<T1 = T>(unions: Array<string>): Blaise<T1>;
  seed<T1 = T>(seed: number): Blaise<T1>;

  getDefault(): BlaiseDefaults<T>;

  eachMessage(
    message?: DeepPartial<AvroKafkaMessage<T>>,
  ): CastleEachMessagePayload<T>;
  eachBatch(messages: Array<AvroKafkaMessage<T>>): CastleEachBatchPayload<T>;
  message(message?: DeepPartial<AvroKafkaMessage<T>>): AvroKafkaMessage<T>;
}

export type BlaiseDefaults<T = unknown> = {
  eachMessage: Omit<CastleEachMessagePayload<T>, 'message'>;
  eachBatch: DeepOmit<
    CastleEachBatchPayload<T>,
    { batch: { messages: never } }
  >;
  message: AvroKafkaMessage<T>;
  avro: AvroOptions & { schema: schema.AvroSchema; seed?: number };
  generator: AvroMock<T>;
};
