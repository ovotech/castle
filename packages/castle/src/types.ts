import {
  AvroProducer,
  AvroEachMessagePayload,
  AvroEachBatchPayload,
  SchemaRegistryConfig,
  AvroConsumer,
  AvroConsumerRun,
  AvroMessage,
  AvroKafka,
  TopicsAlias,
} from '@ovotech/avro-kafkajs';
import {
  ConsumerConfig,
  KafkaConfig,
  ProducerConfig,
  RecordMetadata,
  ConsumerSubscribeTopic,
  KafkaMessage,
} from 'kafkajs';

export interface CastleEachMessagePayload<TValue = unknown, TKey = KafkaMessage['key']>
  extends AvroEachMessagePayload<TValue, TKey> {
  producer: AvroProducer;
}

export interface CastleEachBatchPayload<TValue = unknown, TKey = KafkaMessage['key']>
  extends AvroEachBatchPayload<TValue, TKey> {
  producer: AvroProducer;
}

// eslint-disable-next-line @typescript-eslint/ban-types
export type Resolver<TContext extends {} = {}> = (ctx: TContext) => Promise<void>;

// eslint-disable-next-line @typescript-eslint/ban-types
export type Middleware<TP extends {} = {}, TR extends {} = {}> = <TInherit extends {} = {}>(
  next: Resolver<TP & TR & TInherit>,
) => Resolver<TR & TInherit>;

export interface CastleTopicSubscribeEachMessage<TValue, TKey> {
  eachMessage: (ctx: CastleEachMessagePayload<TValue, TKey>) => Promise<void>;
}

export interface CastleTopicSubscribeEachBatch<TValue, TKey> {
  eachBatch: (ctx: CastleEachBatchPayload<TValue, TKey>) => Promise<void>;
}

export interface CastleTopicSubscribeEachSizedBatch<TValue, TKey> {
  eachSizedBatch: (ctx: CastleEachBatchPayload<TValue, TKey>) => Promise<void>;
  maxBatchSize: number;
}

export type FinalCastleTopicSubscribe<TValue, TKey> =
  | CastleTopicSubscribeEachMessage<TValue, TKey>
  | CastleTopicSubscribeEachBatch<TValue, TKey>;

export type CastleConsumerRun = Omit<AvroConsumerRun, 'eachBatch' | 'eachMessage'>;
export type CastleTopicSubscribe<TValue, TKey> =
  | CastleTopicSubscribeEachMessage<TValue, TKey>
  | CastleTopicSubscribeEachBatch<TValue, TKey>
  | CastleTopicSubscribeEachSizedBatch<TValue, TKey>;

export interface CastleConsumerSubscribeTopic extends Omit<ConsumerSubscribeTopic, 'topic'> {
  topic?: ConsumerSubscribeTopic['topic'];
}

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export type CastleConsumerConfig<TValue = any, TKey = any> = ConsumerConfig &
  CastleConsumerRun &
  ConsumerSubscribeTopic &
  CastleTopicSubscribe<TValue, TKey>;

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export type OptionalCastleConsumerConfig<TValue = any, TKey = any> = ConsumerConfig &
  CastleConsumerRun &
  CastleConsumerSubscribeTopic &
  CastleTopicSubscribe<TValue, TKey>;

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export type FinalCastleConsumerConfig<TValue = any, TKey = any> = ConsumerConfig &
  CastleConsumerRun &
  ConsumerSubscribeTopic &
  FinalCastleTopicSubscribe<TValue, TKey>;

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export interface CastleConsumer<TValue = any, TKey = any> {
  instance: AvroConsumer;
  config: FinalCastleConsumerConfig<TValue, TKey>;
}

export interface CastleConfig {
  kafka: KafkaConfig;
  topicsAlias?: TopicsAlias;
  producer?: ProducerConfig;
  schemaRegistry: SchemaRegistryConfig;
  consumers?: CastleConsumerConfig[];
}

export interface CastleParts {
  kafka: AvroKafka;
  producer: AvroProducer;
  consumers: CastleConsumer[];
}

export interface CastleService {
  connect(): Promise<void>;
  disconnect(): Promise<void>;
}

export interface Castle {
  kafka: AvroKafka;
  consumers: CastleConsumer[];
  producer: AvroProducer;
  isRunning: () => boolean;
  start: () => Promise<void>;
  stop: () => Promise<void>;
}

export type CastleSender<TValue = unknown, TKey = AvroMessage['key']> = (
  producer: AvroProducer,
  messages: AvroMessage<TValue, TKey>[],
) => Promise<RecordMetadata[]>;
