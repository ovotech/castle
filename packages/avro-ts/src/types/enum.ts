import { schema, Schema } from 'avsc';
import { Convert } from '../types';
import { Type } from '@ovotech/ts-compose';
import { namedType, firstUpperCase } from '../helpers';

export const isEnumType = (type: Schema): type is schema.EnumType =>
  typeof type === 'object' && 'type' in type && type.type === 'enum';

export const convertEnumType: Convert<schema.EnumType> = (context, schema) => {
  const namespace = schema.namespace ?? context.namespace;

  const type = Type.Alias({
    name: firstUpperCase(schema.name),
    type: Type.Union(schema.symbols.map((symbol) => Type.Literal(symbol))),
    isExport: true,
    jsDoc: schema.doc,
  });

  return namedType(type, context, schema, namespace);
};
