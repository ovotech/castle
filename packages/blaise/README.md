# Blaise

An API to generate mock payloads for [@ovotech/castle](https://npmjs.com/@ovotech/castle) using [@ovotech/avro-mock-generator](https://npmjs.com/@ovotech/avro-mock-generator)

## Usage

### Concept

`blaise()` is mostly just a chainable deepMerge, with utilities to generate messages and payloads once everything is configured.

Calling `blaise()` returns a new version of blaise, preloaded with the provided defaults, without mutating the parent.

> [examples/chain.ts](examples/chain.ts)

```typescript
import blaise from '@ovotech/blaise'

const withAvro = blaise({ avro: { schema } });
const coffee = withAvro({ avro: { pickUnion: ['coffee'] } });
const tea = withAvro({ avro: { pickUnion: ['tea'] } });

coffee.getDefault(); // { avro: { schema, pickUnion: ['coffee']}}
tea.getDefault(); // { avro: { schema, pickUnion: ['tea']}}
```

### Generate an Avro Message

> [examples/avro-message.ts](examples/avro-message.ts)

```typescript
import blaise from '@ovotech/blaise'

const schema: avsc.RecordType = {
  type: 'record',
  name: '',
  fields: [{ name: 'anInt', type: 'int' }],
};
type MyType = { anInt: number };
blaise<MyType>({ avro: { schema } }).message(); // {topic: '', {anInt: 32}, [...]}

```

You can also override the message, either as part of the default or when calling the function

> [examples/avro-message-override.ts](examples/avro-message-override.ts)

```typescript
import blaise from '@ovotech/blaise'

const schema: avsc.RecordType = {
  // [...]
};
const brewCoffee = blaise({
  avro: { schema },
  message: { value: { type: 'coffee' } },
});
brewCoffee.message({ value: { with: 'milk' } }); // {offset: 13, value {type: 'coffee', with: 'milk'}, [...]}

```

### Generate a `CastleEachMessagePayload`

> [examples/each-message.ts](examples/each-message.ts)

```typescript
import blaise from '@ovotech/blaise'

const schema: avsc.RecordType = {
  // [...]
};
blaise({ avro: { schema }).eachMessage(); // { topic: 'aTopic', value: {} [...]}

```

### Generate a `CastleEachBatchPayload`

> [examples/each-batch.ts](examples/each-batch.ts)

```typescript
import blaise from '@ovotech/blaise'

const schema: avsc.RecordType = {
  // [...]
};
const makeMessage = blaise({ avro: { schema })
blaise.eachBatch([makeMessage()]); // {batch: { topic: 'aTopic', value: {} }[...]}

```

### Deterministic generation

Make use of the `seed` options to use a deterministic random generator instead of true randomness.

Each call to `.defaults()` or `.seed()` with a seed will reset the randomness.

## Running the tests

You can run the tests with:

```bash
yarn test
```

### Coding style (linting, etc) tests

Style is maintained with prettier and eslint

```
yarn lint
```

## Deployment

Deployment is preferment by lerna automatically on merge / push to master, but you'll need to bump the package version numbers yourself. Only updated packages with newer versions will be pushed to the npm registry.

## Contributing

Have a bug? File an issue with a simple example that reproduces this so we can take a look & confirm.

Want to make a change? Submit a PR, explain why it's useful, and make sure you've updated the docs (this file) and the tests.

## License

This project is licensed under Apache 2 - see the [LICENSE](LICENSE) file for details

## What does the name mean?

It's just a fun pun. Take a Castle, make a 'mock' version of it, and you get Bristol's [Blaise Castle](https://en.wikipedia.org/wiki/Blaise_Castle_Estate)
