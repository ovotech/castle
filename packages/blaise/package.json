{
  "name": "@ovotech/blaise",
  "version": "0.2.1",
  "description": "An API to generate mock payloads for @ovotech/castle using @ovotech/avro-mock-generator",
  "keywords": [
    "castle",
    "avro",
    "kafkajs",
    "mock",
    "kafka"
  ],
  "author": "Nicolas Lagoutte <nicolas.lagoutte@ovoenergy.com>",
  "homepage": "https://github.com/ovotech/castle#readme",
  "license": "Apache-2.0",
  "main": "dist/index.js",
  "directories": {
    "dist": "dist",
    "test": "__tests__"
  },
  "files": [
    "dist"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ovotech/castle.git"
  },
  "scripts": {
    "build": "tsc --declaration",
    "build:docs": "build-docs README.md",
    "test": "jest test --runInBand",
    "lint:prettier": "prettier --check src/**/*.ts",
    "lint:eslint": "eslint '{src,test}/**/*.ts'",
    "lint": "yarn lint:prettier && yarn lint:eslint"
  },
  "bugs": {
    "url": "https://github.com/ovotech/castle/issues"
  },
  "jest": {
    "preset": "../../jest.json"
  },
  "devDependencies": {
    "@ovotech/avro-kafkajs": "^0.5.0",
    "@ovotech/castle": "^0.6.0",
    "@types/lodash.merge": "^4.6.6",
    "@typescript-eslint/eslint-plugin": "^3.10.1",
    "@typescript-eslint/parser": "^3.10.1",
    "avsc": "^5.4.21",
    "eslint": "^7.7.0",
    "eslint-config-prettier": "^6.11.0",
    "jest": "^26.4.2",
    "prettier": "^2.1.1",
    "ts-jest": "^26.3.0",
    "typescript": "^4.0.2"
  },
  "dependencies": {
    "@ovotech/avro-mock-generator": "^2.0.3",
    "lodash.merge": "^4.6.2",
    "ts-essentials": "^7.0.0"
  },
  "peerDependencies": {
    "@ovotech/castle": ">=0.3"
  }
}
