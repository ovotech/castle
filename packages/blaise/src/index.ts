import { AvroKafkaMessage } from '@ovotech/avro-kafkajs/dist/types';
import avroMock, { Seeded } from '@ovotech/avro-mock-generator';
import {
  CastleEachBatchPayload,
  CastleEachMessagePayload,
} from '@ovotech/castle';
import merge = require('lodash.merge');
import { DeepPartial } from 'ts-essentials';
import { BlaiseDefaults, Blaise } from './types';

// eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/explicit-function-return-type
const noop = () => {};

/* eslint-disable @typescript-eslint/no-explicit-any */
export const defaultPayload: BlaiseDefaults = {
  eachMessage: {
    topic: 'blaise-default-topic',
    partition: 0,
    producer: noop as any,
  },
  eachBatch: {
    batch: {
      topic: '',
      highWatermark: '',
      partition: 0,
      isEmpty: noop as any,
      firstOffset: noop as any,
      lastOffset: noop as any,
      offsetLag: noop as any,
      offsetLagLow: noop as any,
    },
    isStale: noop as any,
    isRunning: noop as any,
    heartbeat: noop as any,
    uncommittedOffsets: noop as any,
    producer: noop as any,
    resolveOffset: noop as any,
    commitOffsetsIfNecessary: noop as any,
  },
  message: {
    schema: [],
    key: Buffer.from('blaise-default-key'),
    value: {},
    timestamp: Date.now().toString(),
    size: 3,
    attributes: 3,
    offset: '1',
  },
  avro: {
    schema: 'boolean',
  },
  generator: avroMock,
};
/* eslint-enable @typescript-eslint/no-explicit-any */

const buildBlaise = <T = unknown>(defaults = defaultPayload): Blaise<T> => {
  const blaise = <T1 = T>(
    newDefaults?: DeepPartial<BlaiseDefaults>,
  ): Blaise<T1> => {
    if (newDefaults?.avro?.seed) {
      newDefaults.generator = Seeded<T>(newDefaults.avro.seed);
    }
    return buildBlaise<T1>(merge({}, defaults, newDefaults));
  };

  blaise.default = blaise;

  blaise.pickUnion = <T1 = T>(pickUnion: Array<string>) =>
    blaise<T1>({ avro: { pickUnion } });

  blaise.seed = <T1 = T>(seed: number) => blaise<T1>({ avro: { seed } });

  blaise.getDefault = () => defaults as BlaiseDefaults<T>;

  blaise.message = (
    message?: DeepPartial<AvroKafkaMessage<T>>,
  ): AvroKafkaMessage<T> =>
    merge(
      {
        value: defaults.generator(defaults.avro.schema, defaults.avro),
      },
      defaults.message,
      message,
    ) as AvroKafkaMessage<T>;

  blaise.eachMessage = (
    message?: DeepPartial<AvroKafkaMessage<T>>,
  ): CastleEachMessagePayload<T> => ({
    ...defaults.eachMessage,
    message: blaise.message(message),
  });

  blaise.eachBatch = (
    messages: Array<AvroKafkaMessage<T>> = [],
  ): CastleEachBatchPayload<T> =>
    merge({}, defaults.eachBatch, {
      batch: { messages },
    });

  return blaise;
};

export default buildBlaise();
