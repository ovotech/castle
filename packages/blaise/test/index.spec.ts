import blaise, { defaultPayload } from '../src';
import { schema as avsc } from 'avsc/types';

describe('Blaise', () => {
  beforeEach(jest.resetAllMocks);

  it('exposes a function', () => {
    expect(blaise).toBeDefined();
  });

  it('is a function returning default values', () => {
    expect(blaise.getDefault()).toEqual(defaultPayload);
    expect(blaise().getDefault()).toEqual(defaultPayload);
  });

  it('accepts values override', () => {
    expect(
      blaise({ eachMessage: { topic: 'a test' } }).getDefault().eachMessage,
    ).toEqual({
      ...defaultPayload.eachMessage,
      topic: 'a test',
    });
  });

  describe('default()', () => {
    it('deep merges over defaults', () => {
      expect(
        blaise({
          eachMessage: { partition: 1 },
        }).getDefault().eachMessage,
      ).toEqual({
        ...defaultPayload.eachMessage,
        partition: 1,
      });
    });

    it('does not override the previous defaults', () => {
      const blaise1 = blaise({ eachMessage: { topic: '1' } });
      const blaise2 = blaise1({ eachMessage: { topic: '2' } });
      expect(blaise1.getDefault().eachMessage).toEqual(
        expect.objectContaining({ topic: '1' }),
      );
      expect(blaise2.getDefault().eachMessage).toEqual(
        expect.objectContaining({ topic: '2' }),
      );
    });
  });

  describe('With Avro', () => {
    const schema: avsc.RecordType = {
      type: 'record',
      name: '',
      fields: [
        { name: 'added', type: 'string' },
        { name: 'anInt', type: 'int' },
      ],
    };

    const withAvro = blaise<{ added: string; anInt: number; notPass: boolean }>(
      {
        avro: {
          schema,
        },
      },
    );

    describe('message', () => {
      it('is able to generate a random message', () => {
        expect(withAvro.message().value).toEqual(
          expect.objectContaining({
            added: expect.any(String),
            anInt: expect.any(Number),
          }),
        );
      });

      it('allows overriding properties', () => {
        expect(withAvro.message({ value: { added: 'prop' } }).value).toEqual(
          expect.objectContaining({ added: 'prop' }),
        );
      });

      it('respects defaults', () => {
        expect(
          withAvro
            .default({ message: { value: { anInt: 12 } } })
            .message({ value: { added: 'prop' } }).value,
        ).toEqual(expect.objectContaining({ added: 'prop', anInt: 12 }));
      });
    });

    describe('eachMessage', () => {
      it('returns an eachMessagePayload', () => {
        // Test only message and a few other properties. TS has us sorted for the rest
        const payload = withAvro.eachMessage({ value: { added: 'prop' } });
        expect(payload).toEqual(
          expect.objectContaining({
            topic: expect.any(String),
          }),
        );
        expect(payload.message.value).toEqual(
          expect.objectContaining({ added: 'prop' }),
        );
      });
    });

    describe('eachPayload', () => {
      const mockMessage = withAvro.message;

      it('returns an eachMessagePayload', () => {
        // Test only message and a few other properties. TS has us sorted for the rest
        const payload = withAvro.eachBatch([mockMessage()]);
        const message = payload.batch.messages[0];
        expect(payload.batch.topic).toEqual(expect.any(String));
        expect(message.value).toEqual(
          expect.objectContaining({ added: expect.any(String) }),
        );
      });
    });

    describe('pickUnion', () => {
      const schema: avsc.RecordType = {
        type: 'record',
        name: '',
        fields: [
          {
            name: 'onion',
            type: [
              {
                type: 'record',
                name: 'gimmeAString',
                fields: [{ name: 'val', type: 'string' }],
              },
              {
                type: 'record',
                name: 'gimmeAnInt',
                fields: [{ name: 'val', type: 'int' }],
              },
            ],
          },
        ],
      };
      const union = blaise.default({ avro: { schema } });
      it('allows choosing union members', () => {
        expect(union.pickUnion(['gimmeAString']).message()).toHaveProperty(
          'value.onion.gimmeAString.val',
          expect.any(String),
        );
        expect(
          union
            .pickUnion<{ onion: { gimmeAnInt: { val: number } } }>([
              'gimmeAnInt',
            ])
            .message().value.onion.gimmeAnInt.val,
        ).toEqual(expect.any(Number));
      });
    });

    describe('seed', () => {
      it('generates reproducible mocks', () => {
        const seeded = withAvro.default({ avro: { seed: 132 } });
        const {
          message: { value },
        } = seeded.eachMessage();
        expect(value).toMatchSnapshot();
      });

      it('generates different objects by default', () => {
        const seeded = withAvro.seed(132);
        expect(seeded.eachMessage()).not.toMatchObject(seeded.eachMessage());
      });

      it('resets the generator if provided the same seed twice', () => {
        const seeded = withAvro.seed(132);
        const seeded2 = withAvro.seed(132);
        expect(seeded.eachMessage()).toMatchObject(seeded2.eachMessage());
      });

      it("doesn't resets the generator new defaults with the seed unchanged", () => {
        const seeded = withAvro.seed(132);
        const seeded2 = withAvro.default({});
        expect(seeded.eachMessage()).not.toMatchObject(seeded2.eachMessage());
      });
    });
  });
});
