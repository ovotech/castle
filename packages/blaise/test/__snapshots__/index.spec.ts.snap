// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blaise With Avro seed generates reproducible mocks 1`] = `
Object {
  "added": "4b1db2f0-0c24-5e49-8fc8-dc0972d1e8f7",
  "anInt": 1676978352,
}
`;
